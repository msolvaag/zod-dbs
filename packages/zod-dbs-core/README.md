# zod-dbs-core

Shared core functionality for [zod-dbs](https://github.com/msolvaag/zod-dbs).
