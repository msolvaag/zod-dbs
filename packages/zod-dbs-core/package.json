{
  "name": "zod-dbs-core",
  "version": "1.0.2",
  "description": "Shared core functionality for zod-dbs packages",
  "main": "./dist/index.js",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test": "pnpm run test:unit",
    "test:unit": "vitest --project unit run",
    "test:types": "tsc --noEmit"
  },
  "author": "msolvaag",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msolvaag/zod-dbs.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "debug": "^4.4.1"
  }
}
