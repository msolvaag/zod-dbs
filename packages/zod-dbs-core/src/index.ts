export * from './utils/index.js';

export { DatabaseConnector } from './database/DatabaseConnector.js';

export type * from './types.js';
