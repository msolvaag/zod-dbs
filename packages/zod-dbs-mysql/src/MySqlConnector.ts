import { DatabaseConnector, logDebug, sql } from 'zod-dbs-core';

import type {
  ZodDbsConnectionConfig,
  ZodDbsConnectorConfig,
  ZodDbsRawColumnInfo,
} from 'zod-dbs-core';

import { createClient } from './client.js';

interface RawColumnInfo {
  tableName: string;
  name: string;
  ordinalPosition: number;
  defaultValue: string | null;
  isNullable: string;
  dataType: string;
  characterMaximumLength: number | null;
  numericPrecision: number | null;
  numericScale: number | null;
  columnType: string;
  columnKey: string;
  extra: string;
  description: string;
}

/**
 * Default connector to interact with MySQL database and retrieve schema information.
 * Supports MySQL version x and above.
 */
export class MySqlConnector extends DatabaseConnector {
  createClient = (options: ZodDbsConnectionConfig) => {
    return createClient(options);
  };

  public override async fetchSchemaInfo(
    config: ZodDbsConnectorConfig
  ): Promise<ZodDbsRawColumnInfo[]> {
    const { schemaName = 'public' } = config;

    config.onProgress?.('connecting');
    const client = await this.createClient(config);
    await client.connect();

    config.onProgress?.('fetchingSchema');
    logDebug(`Retrieving schema information for schema '${schemaName}'`);

    /**
     * {
    TABLE_CATALOG: 'def',
    TABLE_SCHEMA: 'test',
    TABLE_NAME: 'users',
    COLUMN_NAME: 'id',
    ORDINAL_POSITION: 1,
    COLUMN_DEFAULT: null,
    IS_NULLABLE: 'NO',
    DATA_TYPE: 'bigint',
    CHARACTER_MAXIMUM_LENGTH: null,
    CHARACTER_OCTET_LENGTH: null,
    NUMERIC_PRECISION: 20,
    NUMERIC_SCALE: 0,
    DATETIME_PRECISION: null,
    CHARACTER_SET_NAME: null,
    COLLATION_NAME: null,
    COLUMN_TYPE: 'bigint unsigned',
    COLUMN_KEY: 'PRI',
    EXTRA: 'auto_increment',
    PRIVILEGES: 'select,insert,update,references',
    COLUMN_COMMENT: '',
    GENERATION_EXPRESSION: '',
    SRS_ID: null
  },
     */
    try {
      const result = await client.query<RawColumnInfo[]>(
        sql`
          SELECT
            TABLE_NAME AS "tableName",
            COLUMN_NAME AS "name",
            ORDINAL_POSITION AS "ordinalPosition",
            COLUMN_DEFAULT AS "defaultValue",
            IS_NULLABLE AS "isNullable",
            DATA_TYPE AS "dataType",
            CHARACTER_MAXIMUM_LENGTH AS "characterMaximumLength",
            NUMERIC_PRECISION AS "numericPrecision",
            NUMERIC_SCALE AS "numericScale",
            COLUMN_TYPE AS "columnType",
            COLUMN_KEY AS "columnKey",
            EXTRA AS extra,
            COLUMN_COMMENT AS "description"
          FROM information_schema.columns
          WHERE TABLE_SCHEMA = ?
        `,
        [schemaName]
      );

      return result.map(
        (row): ZodDbsRawColumnInfo => ({
          tableName: row.tableName,
          name: row.name,
          isNullable: row.isNullable === 'YES',
          dataType: row.dataType,
          defaultValue: row.defaultValue,
          maxLen: row.characterMaximumLength ?? null,
          schemaName,
          tableType: 'table',
        })
      );
    } finally {
      await client.end();
    }
  }
}
