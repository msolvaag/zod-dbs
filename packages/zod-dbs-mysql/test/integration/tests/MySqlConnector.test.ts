import { ZodDbsConnectionConfig } from 'zod-dbs-core';

import { MySqlConnector } from '../../../src/MySqlConnector.js';
import {
  getConnectionConfig,
  setupTestDb,
  teardownTestDb,
  TestDbContext,
} from '../testDbUtils.js';

let ctx: TestDbContext;
let connectionConfig: ZodDbsConnectionConfig;

beforeAll(async () => {
  ctx = await setupTestDb();
  connectionConfig = getConnectionConfig();
});

afterAll(async () => {
  await teardownTestDb(ctx);
});

it.only('returns raw schema column information', async () => {
  const connector = new MySqlConnector();

  const info = await connector.fetchSchemaInfo(connectionConfig);

  expect(info).toBeDefined();
  expect(info).toHaveLength(6);
  expect(info).toMatchSnapshot('rawColumns');
});

it('returns schema models', async () => {
  const connector = new MySqlConnector();

  const info = await connector.getSchemaInformation(connectionConfig);

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();
  expect(info.tables).toHaveLength(1);

  const userTable = info.tables.find((t) => t.name === 'users')!;
  expect(userTable).toBeDefined();
  expect(userTable.columns).toBeDefined();
  expect(userTable.columns).toHaveLength(8);

  expect(userTable).toMatchSnapshot('userTable');
});
