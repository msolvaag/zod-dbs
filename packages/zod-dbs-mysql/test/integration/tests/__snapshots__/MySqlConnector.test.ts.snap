// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns raw schema column information > rawColumns 1`] = `
[
  {
    "dataType": "timestamp",
    "defaultValue": "CURRENT_TIMESTAMP",
    "isNullable": true,
    "maxLen": null,
    "name": "created_at",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": null,
    "isNullable": true,
    "maxLen": 255,
    "name": "email",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "bigint",
    "defaultValue": null,
    "isNullable": false,
    "maxLen": null,
    "name": "id",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": null,
    "isNullable": false,
    "maxLen": 100,
    "name": "name",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "json",
    "defaultValue": null,
    "isNullable": true,
    "maxLen": null,
    "name": "profile",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": "active",
    "isNullable": true,
    "maxLen": 20,
    "name": "status",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
]
`;
